### DISCUSSION TITLE
Relay not getting triggered

### DISCUSSION THREAD

**Discussion by ThunderBirrddd** [HomeSpan/HomeSpan#677] [Q&A]
i have used homespan on a esp32-wroom to create a single relay smart homekit device 

everything is working perfectly fine the device got registered added in the homekit flawlessly 
but the issue is whenever im trying to turn the switch on and off the relay doesnt tigger and stay on the (on) state. 

ill provide my code and my wiring diagram aswell .
please help.

this is the main tab.

```
`#include "HomeSpan.h"
#include "DEV_RELAY.h"

void setup() {
  Serial.begin(115200);
  homeSpan.setPairingCode("11122333");
  homeSpan.begin(Category::Bridges, "Ambient Light");

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify();
    new DEV_RELAY(17);
}

void loop() {
  homeSpan.poll();

}

```
next ive created another tab to organize my code.

```
struct DEV_RELAY : Service::Switch {
  
  int relayPin;
  SpanCharacteristic *power;

  DEV_RELAY(int relayPin) : Service::Switch(){

      power = new Characteristic::On();
      this->relayPin = relayPin;
      pinMode(relayPin, OUTPUT); 
    }
  

  boolean update() {
    
      digitalWrite(relayPin, power->getNewVal());

      return(true);
    }
  };

```

here is the wiring diagram.
![power](https://github.com/HomeSpan/HomeSpan/assets/39340381/8e431386-3b4f-4c91-bedd-1b126d5b6f4d)

**Comment by HomeSpan** [comment]
To debug, I recommend disconnecting the relay and using GPIO 17 to drive a simple LED (through a resistor if needed).  Then check that the LED turns on/off correctly when you control with the Home App.  If that works as expected, you know the code is fine and that the issue is either in the wiring or the relay.

**Reply by ThunderBirrddd** [reply]
How do you wire a led through gpio ? 
Leds have 2 pins negative and positive where does the gpio go ?

**Comment by HomeSpan** [comment]
GPIO goes to resistor and then positive side of LED.  Negative goes to GND.

I looked up the relay you have on the web and it likely won't work - it requires 5V to drive but the GPIOs on the ESP32 are 3.3V.  This is likely the cause of the problem.  There are lots of articles out there that describe how to use relays, leds, etc. with an ESP32.  Based on the relay that you chose, it looks like you are trying to switch something connected to normal house voltage.  If you've not done things like this before (I'm assuming not based on your LED question), I highly recommend you first experiment with low-voltage only.  As noted, there are lots of tutorials on the web you can use that are unrelated to HomeSpan but would give you all the foundations.   Switching house voltage can be very dangerous without the proper experience.

**Reply by ThunderBirrddd** [reply]
So I need a 5V line connection to power it as well as a 5V IN connection which triggers it ?

Yes I’m kinda new to the smart wifi enabled diy but I’ve worked with relays before.

**Comment by HomeSpan** [comment]
It depends on the relay.  Some relays may be able to operate at a lower voltage level.  I did a quick web search and it seems this relay is problematic if you try to operate at a lower voltage.  There are ways to address this - for example, you can drive with a transistor instead of directly with a GPIO.

**Reply by ThunderBirrddd** [reply]
So I have connected 3 ports to relay as seen in the wiring diagram 

Relay:       ESP32
IN      -      GPIO17
VCC   -       5V pin 
Gnd    -      Ground 

Now if u see I’m already providing 5Volts to the relay

**Reply by HomeSpan** [reply]
Yes, but the IN is only getting 3.3 volts.  You need to find the spec sheet for the relay to determine whether or not 3.3V on the IN pin will operate the relay.

**Reply by ThunderBirrddd** [reply]
i tried wiring a led and yes it did work.

But the relay doesnt trigger,

now to increase the voltage of GPIO17 from 3.3v to 5v what should i do ?

**Reply by ThunderBirrddd** [reply]
and does this mean the relay requires 2 - 5v inputs to trigger?

**Reply by HomeSpan** [reply]
I'm unfamiliar with the details of your relay so I can't really provide any further advice on how to connect it to an ESP32.  You may want to purchase a relay one of the popular hobby sites - they usually come with detailed tutorials on how to connect and operate from an ESP32.

**Reply by daadaan** [reply]
I am facing the same issue. I have replaced the abovementioned relay with an optocoupler relay (5V). The problem is that the switching action is reversed - when I turn the switch off in the Home app, the relay registers as ON state. I think the assigned GPIO pin of the ESP32  gets power when in OFF state and vice versa. 

My code is same as above. Is there something we're doing wrong?

**Comment by daadaan** [comment]
I faced the same problem with the same relay and board. I tested a 5V optocoupler relay and the wiring seemed reversed, i.e. the relay switched on when I switched off from the Home app. Here is a quick patch for the problem and does work. Somehow the GPIO pin switches ON if the homekit switch is in OFF state and vice versa, therefore I have inversed the logic.

Keep in mind the relay shown in the picture does not work, it probably needs a 5V control voltage (I'm speculating).

```
struct DEV_RELAY: Service::LightBulb {
    int relayPin;
    bool invert;
    SpanCharacteristic *state;
    
    DEV_RELAY(int pin, bool defaultOn = false, bool invert = true)
    : Service::LightBulb()
    , invert(invert) {
        this->state = new Characteristic::On(defaultOn);
        this->relayPin = pin;
        pinMode(this->relayPin, OUTPUT);
        digitalWrite(this->relayPin, 1);
    }
    
    boolean update() {
        digitalWrite(
            this->relayPin,
            this->invert ^ this->state->getNewVal()
        );
        return (true);
    }  
};
```

**Reply by krishnafujin** [reply]
Thank you @daadaan  works like a  charm after inversing logic.

**Comment by YufransAdiMakita** [comment]
Hay there, 
![Uploading image.jpg…]()
I’ve the same problem so i consult with my teacher and i solve that problem, just add “!” Before power on boolean update and it solve
